### DISCUSSION TITLE
Chrome extension apis - manifest v2 vs. v3

### DISCUSSION THREAD

**Discussion by yureshwar** [Digital-Assistant/Digital_Assistant_Client#329] [Show and tell]
List of chrome apis used in UDAN chrome extension

-  background
-  storage
-  identity
-  identity.email
-  webRequest
-  webRequestBlocking
-  oauth2
-  chrome.tabs
-  chrome.storage
-  localStorage
-  chrome.webRequest.onHeadersReceived
-  web_accessible_resources

**Comment by yureshwar** [comment]
In v3 remotely hosted code is getting blocked hence our approach of finding clickable objects will no longer work in v3

"Google is right to ban remotely hosted code (with some exceptions for things like user scripts), but this is a policy change that didn’t need to be bundled with the rest of"

https://www.eff.org/deeplinks/2021/12/googles-manifest-v3-still-hurts-privacy-security-innovation

**Comment by yureshwar** [comment]
Manifest V3 removes the choice, instead of requiring all extensions to be based on “service workers.” Service workers are ephemeral, event-based, and do not have access to the standard set of website APIs. Along with removing the “blocking webRequest”

This will cause an issue while processing the node elements. With v3, node elements cannot be sent to service workers.

https://www.eff.org/deeplinks/2021/12/googles-manifest-v3-still-hurts-privacy-security-innovation

**Comment by kcramakrishna** [comment]
1. We need to double check that V3 will not support our extension. Need to check if Chrome makes it available to Enterprises.

2. Microsoft/Edge too are embracing V3 - https://docs.microsoft.com/en-us/microsoft-edge/extensions-chromium/developer-guide/manifest-v3 BUT Edge may decide to accommodate large Enterprises.

3. Firefox may decide to maintain backward compatibility. But this needs to be validated. **"Mozilla will maintain support for blocking WebRequest in MV3."**  https://blog.mozilla.org/addons/2022/05/18/manifest-v3-in-firefox-recap-next-steps/ . Currently this is our best bet.
We can try implementing on FF with v3 and see if it works. https://extensionworkshop.com/documentation/develop/manifest-v3-migration-guide/

5. We need to explore how this will work with Brave etc. - At least publicly, Brave etc have decided to support the old manifest. https://www.zdnet.com/article/opera-brave-vivaldi-to-ignore-chromes-anti-ad-blocker-changes-despite-shared-codebase/ 

6. We also need to explore whether we can do the same with Safari.

**Comment by kcramakrishna** [comment]
More info in this thread: https://groups.google.com/a/chromium.org/g/chromium-extensions/c/wczJuX04RrM/m/xoAfxIMVAQAJ

There seems to be some progress here. We have to experiment with this: https://developer.chrome.com/docs/extensions/whatsnew/#m95-page-script-injection

**Comment by yureshwar** [comment]
There is a big change in background.js and we may have to use workaround for connecting the clickable nodes to the UDAN sdk

![image](https://user-images.githubusercontent.com/516516/175505941-98724aa7-8f85-43d2-87fa-a7c5366ee47e.png)

https://developer.chrome.com/docs/extensions/mv3/intro/mv3-migration/#background-service-workers

**Comment by yureshwar** [comment]
We might achieve the manifest v2 functionality using "executing arbitrary strings"

![image](https://user-images.githubusercontent.com/516516/175508437-a712e4ec-d257-4014-a981-7f89d8fc5083.png)

https://developer.chrome.com/docs/extensions/mv3/intro/mv3-migration/#executing-arbitrary-strings

**Comment by yureshwar** [comment]
V3 is compatible for our extension and we can inject the scripts using "content_scripts" and by specifying option "run_at" in the content scripts

https://developer.chrome.com/docs/extensions/mv3/content_scripts/#run_time
